### PR TITLE
gh-99108: Initial import of HACL-SHA3 into Python

### PR DESCRIPTION
**Pull request by msprotz**
I have a PR that's close to landing upstream (https://github.com/hacl-star/hacl-star/pull/803). If there are any comments here, I'll be happy to tweak the PR upstream.

This replaces the in-tree SHA3 implementation for hashlib with a verified version from the HACL* project.

A few remarks:
- just like the original implementation, the HACL implementation uses a single API for all six Keccak variants
- I did not understand why there was a lock in sha3module.c, along with critical sections (ENTER_HASHLIB, etc.) when none of the other hashing modules have them -- I have removed this part

@gpshead please let me know what you think! I'm around, and can help make sure this gets into 3.12

* Issue: gh-99108

### CODE REVIEW

**Comment by msprotz** [comment]
I've refreshed this PR with a newer upstream.
- Eliminates an unused variable.
- Switches to a saner API of finish (SHA3) and squeeze (extra length argument, only for Shake).

@gpshead the upstream PR will land soon, let me know if there's anything I can do to help make sure this gets into 3.12!

**Comment by msprotz** [comment]
I've landed the upstream PR, meaning that this is now pulling from HACL*'s main branch. I see from https://peps.python.org/pep-0693/ that the first 3.12 beta is in just a few days. I'll keep an eye on this PR in case any change or fixup is needed. Thanks!

**Comment by gpshead** [comment]
At first glance I believe this looks good, I'll get it in.  I'm running it through the buildbots to see if anything unusual build-wise falls out to be addressed.  _(expect some flaky infrastructure or pre-existing issue noise among those results as well)_

**Comment by msprotz** [comment]
Thanks, so far only the refleaks test seems relevant -- it says hashlib leaked. Is there any way to get a trace of the allocations that weren't freed? Nothing jumps out at me looking at the code right now, and I don't believe anything was done differently compared to the other hashes (for which there were no leaks issues).

**Comment by msprotz** [comment]
I merged master in and I think I found the missing PyBuffer_Release. Can you run buildbot again?

**Comment by msprotz** [comment]
This is the only error that shows up (WASI build):

```
ERROR: test_with_segments (test.test_pathlib.PathSubclassTest.test_with_segments)
ERROR: test_with_segments (test.test_pathlib.PathTest.test_with_segments)
ERROR: test_with_segments (test.test_pathlib.PosixPathTest.test_with_segments)
```

This doesn't seem to be related, does it?

File: PCbuild/pythoncore.vcxproj [hunk 1]
```diff
@@ -381,7 +381,7 @@
     <ClCompile Include="..\Modules\_lsprof.c" />
     <ClCompile Include="..\Modules\_pickle.c" />
     <ClCompile Include="..\Modules\_randommodule.c" />
-    <ClCompile Include="..\Modules\_sha3\sha3module.c" />
+    <ClCompile Include="..\Modules\sha3module.c" />
```
Comment (AA-Turner) [inline PCbuild/pythoncore.vcxproj]: Should this be moved to keep the list alphabetical? (pt. 1)

```suggestion
```

File: PCbuild/pythoncore.vcxproj [hunk 2]
```diff
@@ -415,6 +415,7 @@
     <ClCompile Include="..\Modules\sha1module.c" />
     <ClCompile Include="..\Modules\_hacl\Hacl_Streaming_SHA2.c" />
     <ClCompile Include="..\Modules\sha2module.c" />
+    <ClCompile Include="..\Modules\_hacl\Hacl_Hash_SHA3.c" />
```
Comment (AA-Turner) [inline PCbuild/pythoncore.vcxproj]: Should this be moved to keep the list alphabetical? (pt. 2)
```suggestion
    <ClCompile Include="..\Modules\_hacl\Hacl_Hash_SHA3.c" />
    <ClCompile Include="..\Modules\sha3module.c" />
```

File: PCbuild/pythoncore.vcxproj.filters [hunk 3]
```diff
@@ -794,7 +794,7 @@
     <ClCompile Include="..\Modules\_randommodule.c">
       <Filter>Modules</Filter>
     </ClCompile>
-    <ClCompile Include="..\Modules\_sha3\sha3module.c">
+    <ClCompile Include="..\Modules\sha3module.c">
```
Comment (AA-Turner) [inline PCbuild/pythoncore.vcxproj.filters]: Same comment on lexicographical order.

**Comment by msprotz** [comment]
@AA-Turner I pushed a commit to enforce better ordering.

I'm not sure how the list is ordered, though: from _abc.c to _weakref.c are files starting with an underscore, correctly sorted. Then, there is a mix of underscore-prefixed files and non-underscore-prefixed files, sorted by alphabetical ordering, after removing the initial underscore: `arraymodule.c`, ..., `_operator.c`, `posixmodule.c`, and so on.

I did my best to try to make the list look sensible. I hope this addresses your comments, let me know if it doesn't. Thanks!

**Comment by AA-Turner** [comment]
@zooba would likely be the best to comment on what *should* be the case, though your changes seem alright to me.

A

**Comment by msprotz** [comment]
Great, thanks. I'm also happy to do a followup PR to fix things up after this one lands, if need be.

**Review by gpshead — APPROVED** [review]

File: Modules/Setup [hunk 4]
```diff
@@ -166,7 +166,7 @@ PYTHONPATH=$(COREPYTHONPATH)
 #_md5 md5module.c -I$(srcdir)/Modules/_hacl/include _hacl/Hacl_Hash_MD5.c -D_BSD_SOURCE -D_DEFAULT_SOURCE
 #_sha1 sha1module.c -I$(srcdir)/Modules/_hacl/include _hacl/Hacl_Hash_SHA1.c -D_BSD_SOURCE -D_DEFAULT_SOURCE
 #_sha2 sha2module.c -I$(srcdir)/Modules/_hacl/include Modules/_hacl/libHacl_Streaming_SHA2.a
-#_sha3 _sha3/sha3module.c
+#_sha3 sha3module.c -I$(srcdir)/Modules/_hacl/include _hacl/Hacl_Hash_sha3.c -D_BSD_SOURCE -D_DEFAULT_SOURCE
```
Comment (asottile) [inline Modules/Setup:169]: this line has a casing mistake which breaks case sensitive filesystems

**Pull request merged**